feat: added support for AsyncAPI v3

nashjain opened the pull request:
- Added support for AsyncAPI v3
- Also cleaned up the spec to make it very clear that step-object can be oneOf openapi-step-object or asyncapi-step-object or workflow-step-object 
- For AsyncAPI we really need support for timeout, fork and join. However, these are also useful for OpenAPI so added it at the base step object. 
- For OpenAPI we need at least one successCriteria but for AsyncAPI it can be optional.

While we've tried to incorporate as much as possible from #270 not everything is covered.

frankkilcommins commented:
Thanks @nashjain - I will try to propose changes to the specification based on the examples later this week. There will no doubt be a few finer grained points that we'll need to discuss

nashjain commented:
Thanks @frankkilcommins The proposal mostly looks good to me. I just need sometime to think through a couple of items. I'm currently in Australia. Next week, once I'm back we could jump on a call to discuss and close it.

nashjain committed:
Added support for AsyncAPI v3
Also cleaned up the spec to make it very clear that step-object can be oneOf openapi-step-object or asyncapi-step-object or workflow-step-object
For AsyncAPI we really need support for timeout and dependOn. However, these are also useful for OpenAPI/Workflow steps so added it at the base step object.
For OpenAPI we need at least one successCriteria but for AsyncAPI it can be optional.

nashjain committed:
Update arazzo.md file with AsyncAPI related changes along with a sample AsyncAPI specification for Pet Purchase API

DmitryAnansky commented on src/arazzo.md:
```diff
@@ -173,11 +178,35 @@ workflows:
-      # outputs from this step
-      availablePets: $response.body
+      petId: $response.body#/0/id
+  - stepId: purchasePetStep
```
Could you please update the operationId in the first step.
```
If multiple (non arazzo type) sourceDescriptions are defined, then the operationId MUST be specified using a [Runtime Expression](https://spec.openapis.org/arazzo/latest.html#runtime-expressions) (e.g., $sourceDescriptions.<name>.<operationId>) to avoid ambiguity or potential clashes.
```

nashjain replied:
Thanks. This has been fixed now.

DmitryAnansky commented on src/arazzo.md:
```diff
@@ -296,19 +325,24 @@ Describes a single workflow step which MAY be a call to an API operation ([OpenA
+| <a name="stepTimeout"></a>timeout                 |                                        `integer`                                         | The maximum number of milli-seconds to wait for the step to complete before aborting and failing the step. Consequently this will fail the workflow unless failureActions are defined.                                                                         ...
+| <a name="stepCorrelationId"></a>correlationId     |                                         `string`                                         | A correlationId in AsyncAPI links a request with its response (or more broadly, to trace a single logical transaction across multiple asynchronous messages). Only applicable to `asyncapi` steps with action `receive` and has to be in-sync with correlation ...
+| <a name="stepAction"></a>action                   |                                    `send or receive`                                     | Describes the intent of the message flow. Indicates whether the step will send (publish) or receive (subscribe) to a message on a channel described in an AsyncAPI document, Only applicable for `asyncapi` steps.                                             ...
+| <a name="stepDependsOn"></a>dependsOn             |                                      List[`string`]                                      | A list of steps that MUST be completed before this step can be executed. Each value provided MUST be a `stepdId`. The keys must follow the regular expression: `^\$steps.<stepId>$`.                                                                           ...
```
Does this mean that Step Object DependsOn can be only defined with the step inside `current` workflow?

DmitryAnansky commented on src/arazzo.md:
```diff
@@ -377,6 +411,32 @@ steps:
+- stepId: confirmOrder
+  description: This step demonstrates the action of receiving a message payload to confirm an order
+  operationId: $sourceDescriptions.asyncOrderApi.confirmOrder
+  correlationId: $steps.placeOrder.outputs.requestId
```
Could you please define outputs in `stepId: placeOrder` before using it in `stepId: confirmOrder`

nashjain replied:
Thanks. This has been fixed now. We are directly ref to the input and we don't need to depend on outputs.

DmitryAnansky commented on src/arazzo.md:
```diff
@@ -296,19 +325,24 @@ Describes a single workflow step which MAY be a call to an API operation ([OpenA
+| <a name="stepRequestBody"></a>requestBody         |                       [Request Body Object](#request-body-object)                        | The request body to pass to an operation as referenced by `operationId` or `operationPath`. The `requestBody` is fully supported in HTTP methods where the HTTP 1.1 specification [RFC9110](https://tools.ietf.org/html/rfc9110#section-9.3) explicitly define ...
+| <a name="stepSuccessCriteria"></a>successCriteria |                         [[Criterion Object](#criterion-object)]                          | A list of assertions to determine the success of the step. Each assertion is described using a [Criterion Object](#criterion-object). All assertions `MUST` be satisfied for the step to be deemed successful.                                                 ...
+| <a name="stepOnSuccess"></a>onSuccess             | [[Success Action Object](#success-action-object) \| [Reusable Object](#reusable-object)] | An array of success action objects that specify what to do upon step success. If omitted, the next sequential step shall be executed as the default behavior. If multiple success actions have similar `criteria`, the first sequential action matching the cr ...
+| <a name="stepOnFailure"></a>onFailure             | [[Failure Action Object](#failure-action-object) \| [Reusable Object](#reusable-object)] | An array of failure action objects that specify what to do upon step failure. If omitted, the default behavior is to break and return. If multiple failure actions have similar `criteria`, the first sequential action matching the criteria SHALL be the act ...
```
I am wondering if you already discussed how asyncApi steps should behave with `retry` on Failure actions.
Will it subscribe again, or do any other actions? How to handle potentially duplicated messages in queues?
Maybe I am missing smth.

nashjain committed:
fix: update operationId and correlationId references in arazzo.md AND add pattern for dependsOn in schema.yaml to clarify that it can ref to a step from another workflow

frankkilcommins reviewed (commented):
@nashjain can you please update the PR to target `v1.1-dev` branch? Thanks

kevinduffey reviewed (dismissed):
Overall looks fantastic. We've discussed a bunch and my only request at this stage is that we consider adding some sort of test details for future tool providers to utilize in ensuring their tool supports the addition of async.

frankkilcommins commented on src/arazzo.md:
```diff
@@ -296,19 +325,24 @@ Describes a single workflow step which MAY be a call to an API operation ([OpenA
+| <a name="stepTimeout"></a>timeout                 |                                        `integer`                                         | The maximum number of milli-seconds to wait for the step to complete before aborting and failing the step. Consequently this will fail the workflow unless failureActions are defined.                                                                         ...
+| <a name="stepCorrelationId"></a>correlationId     |                                         `string`                                         | A correlationId in AsyncAPI links a request with its response (or more broadly, to trace a single logical transaction across multiple asynchronous messages). Only applicable to `asyncapi` steps with action `receive` and has to be in-sync with correlation ...
+| <a name="stepAction"></a>action                   |                                    `send or receive`                                     | Describes the intent of the message flow. Indicates whether the step will send (publish) or receive (subscribe) to a message on a channel described in an AsyncAPI document, Only applicable for `asyncapi` steps.                                             ...
+| <a name="stepDependsOn"></a>dependsOn             |                                      List[`string`]                                      | A list of steps that MUST be completed before this step can be executed. Each value provided MUST be a `stepdId`. If you depend on a step from another workflow, you MUST use the full reference format with the workflow id. The keys MUST follow the regular ...
```
@tatomyr - thanks for raising this. It's an important point that we could clarify within the spec.

IMO, Introducing step-level `dependsOn` does not require parallel execution, but it does enable it. 

Steps without `dependsOn` MAY be considered independent and thus MAY begin execution immediately, including in parallel, if tooling supports it.

Tools that supports only sequential execution MUST fall back to topologically sorted execution of steps, respecting all declared `dependsOn` relationships are respected.

It would be beneficial to include additional authoring guidance, something along the lines of:

> If any step defines a `dependsOn`, authors SHOULD declare dependencies for _all steps_ where execution order matters. Steps without `dependsOn` or implicit **output references** MAY be interpreted as suitable for parallel execution. Do not rely on the order of the `steps` array when using `dependsOn`.

The alternative to this would be to constrain `dependsOn` usage to asynchronous operations (akin to an `await` semantics). While that simplifies the model, it limits the utility of `dependsOn` for broader use cases that would benefit from more general dependency-driven execution, including parallelism and fan-in/fan-out coordination.

tatomyr replied:
@frankkilcommins wouldn't this be a breaking change? 

>  dependsOn does not require parallel execution, but it does enable it.

If we allow that, it can break existing workflows, as it introduces ambiguity in step sequencing that don't have the `dependsOn` relation. E.g., a workflow describing some CRUD sequence can potentially be executed simultaneously (unless `dependsOn` is explicitly added for each step), which might not be what the user expects.

Also, this looks to me overly complicated and error-prone, because it becomes harder to write, easier to miss something, and way harder to read, as now you'll have to do mental gymnastics to re-sort the steps. 
Even in [the example provided](https://github.com/OAI/Arazzo-Specification/pull/367#discussion_r2627978295), the steps `createOrderReceive` and `createOrderSend` could potentially be executed in parallel (because there's no `dependsOn` declared between them), but it doesn't make much sense since the first does depend on the  latter (so I guess it was supposed to have the `dependsOn` relation, but that was simply missed).

Again, what is the need for steps parallelism for AsyncAPI operations in the first place? If we do want parallelism, it can be achieved via different workflows.

kevinduffey reviewed (dismissed):
LGTM

nashjain committed:
fix: enhance dependsOn description in schema.yaml and arazzo.md, and add new AsyncAPI examples for test

nashjain committed:
fix: update stepId in examples

spirefyio reviewed (approved):
Get it done!

kevinduffey reviewed (dismissed):
LGTM

DmitryAnansky commented on tests/schema/pass/openapi-and-asyncapi.arazzo.yaml:
```diff
@@ -0,0 +1,129 @@
+    operationId: "$sourceDescriptions.AsyncOrderApi.outForDelivery"
+    action: "send"
+    dependsOn:
+    - "$steps.createOrderReceive"
```
Do we really need to reference `$steps.createOrderReceive`?
Since stepId is already unique, maybe it would make sense to use just `createOrderReceive`.

DmitryAnansky commented on tests/schema/pass/openapi-and-asyncapi.arazzo.yaml:
```diff
@@ -0,0 +1,129 @@
+  - stepId: "reserveInventory"
+    operationId: "$sourceDescriptions.WarehouseApi.reserveInventory"
+    dependsOn:
+    - "$steps.createOrderReceive"
```
It would be nice to have this case covered in unit tests as well.

frankkilcommins replied:
ambiguity around expressions is addressed in #454 - resolving this comment thread

nashjain committed:
Apply suggestion from @frankkilcommins

Co-authored-by: Frank Kilcommins <fkilcommins@gmail.com>

nashjain committed:
refactor: update dependsOn references to remove $steps prefix and clarify usage in schema

nashjain committed:
fix: remove quotes from YAML runtime expressions and dependsOn references

frankkilcommins reviewed (dismissed):
Thanks for all the work on this @nashjain. We might tidy up examples to remove redundant quotes etc. but that can happen in a small house-keeping PR as we prepare information for the learn site.

RomanHotsiy commented on src/arazzo.md:
```diff
@@ -379,6 +413,32 @@ steps:
+  operationId: $sourceDescriptions.asyncOrderApi.confirmOrder
+  correlationId: $inputs.correlationId
+  action: receive
+  dependsOn: placeOrder
```
there are more examples like this throughout this PR

RomanHotsiy replied:
all fixed now, thx

nashjain committed:
fix: update dependsOn format to use a list in arazzo.md

nashjain committed:
fix: remove quotes from YAML keys and update formatting in openapi-and-asyncapi.arazzo.yaml

nashjain commented on src/schemas/validation/schema.yaml:
```diff
@@ -240,21 +263,94 @@ $defs:
+            description: A list of assertions to determine the success of the step
+            type: array
+            uniqueItems: true
+            minItems: 0
```
If I'm firing a notification (which is typically a fire-n-forget), can you please help me understand what kind of things you would check in the successCriteria?

It would really help if you can share a concrete real-world example, so we can resolve this question.

nashjain commented on tests/schema/pass/openapi-and-asyncapi.arazzo.yaml:
```diff
@@ -0,0 +1,129 @@
+    operationId: $sourceDescriptions.AsyncOrderApi.orderAccepted
+    action: receive
+    correlationId: $steps.createOrderSend.outputs.requestId
+    timeout: 6000
```
See the below comment.

AlbinaBlazhko17 commented on tests/schema/pass/async-depends-on.arazzo.yaml:
```diff
@@ -0,0 +1,39 @@
+        operationId: deleteUser
+        action: send
+        dependsOn:
+          - $workflows.dependsOnWorkflow.steps.step2
```
But there are no other workflows in this file. `step4` is inside the `dependsOnWorkflow` workflow, in the same way that `step3` depends on just `step2`.

nashjain committed:
refactor: move default successCriteria to workflow-step from step-object-base to avoid confusion. Add success criteria examples for AsyncAPI, OpenAPI and workflow steps

nashjain committed:
refactor: move successCriteria back to step-object-base. If successCriteria key is present, then at least one criterion-object should be present.

nashjain commented on tests/schema/pass/async-depends-on.arazzo.yaml:
```diff
@@ -0,0 +1,39 @@
+      - stepId: step2
+        operationId: fetchUserId
+        action: receive
+        correlationId: id123
```
You are right from a demonstration of how real correlationId works, correlationId should be sourced from step1 or from the real source of the id. 

However, please note, as the title suggests, the purpose of this example is to check schema validity of just dependsOn. These are internal author unit tests, not end customer training examples.

AlbinaBlazhko17 commented on src/schemas/validation/schema.yaml:
```diff
@@ -240,21 +256,87 @@ $defs:
+      - $ref: '#/$defs/step-object-base'
+      - type: object
+        properties:
+          workflowId:
```
Could you please clarify this? Looking at the schema, `workflow-step-object` inherits `operationId` from `step-object-base`, but there is no constraint enforcing that `workflowId` and `operationId` are mutually exclusive. The spec said, that they are mutually exclusive, but the schema currently allows declaring both simultaneously without any validation error.

nashjain committed:
test: add invalid workflow step example with operationId restriction

nashjain committed:
fix: clarify description of dependsOn field in schema.yaml and update formatting in arazzo.md

nashjain commented on tests/schema/pass/async-depends-on-across-document.arazzo.yaml:
```diff
@@ -0,0 +1,29 @@
+        operationId: localSend
+        action: send
+        dependsOn:
+          - waitForLocalEvent
```
I'm guessing you are suggesting that we should ref to a workflow id instead of a specific step inside the workflow.

My concern with referring only to the workflow id is that it makes the dependency too coarse-grained. A workflow can have multiple async steps, and another workflow may need to proceed when the first async step completes, rather than waiting for the whole workflow to finish. So if step-level cross-workflow dependencies are a valid use case, we may need a way to reference a specific async step, not just the workflow as a whole.

RomanHotsiy commented on src/arazzo.md:
```diff
@@ -298,19 +327,24 @@ Describes a single workflow step which MAY be a call to an API operation ([OpenA
+| <a name="stepTimeout"></a>timeout                 |                                        `integer`                                         | The maximum number of milli-seconds to wait for the step to complete before aborting and failing the step. Consequently this will fail the workflow unless failureActions are defined.                                                                         ...
+| <a name="stepCorrelationId"></a>correlationId     |                                         `string`                                         | A correlationId in AsyncAPI links a request with its response (or more broadly, to trace a single logical transaction across multiple asynchronous messages). Only applicable to `asyncapi` steps with action `receive` and has to be in-sync with correlation ...
+| <a name="stepAction"></a>action                   |                                    `send or receive`                                     | Describes the intent of the message flow. Indicates whether the step will send (publish) or receive (subscribe) to a message on a channel described in an AsyncAPI document, Only applicable for `asyncapi` steps.                                             ...
+| <a name="stepDependsOn"></a>dependsOn             |                                      List[`string`]                                      | A list of steps that MUST be completed before this step can be executed. Each value provided MUST be a `stepId`. The `stepId` value is case-sensitive. If the step depended on is defined within the **current workflow**, specify the `stepId` directly (e.g. ...
```
> That said, if people feel this could be interpreted inconsistently, I would request you to share a lightweight clarification or example rather than a strict definition.

Well, I don't understand it 😄 . Can OpenAPI step be async/non-blocking? or only AsyncAPI step?

AlbinaBlazhko17 commented on tests/schema/pass/async-depends-on.arazzo.yaml:
```diff
@@ -0,0 +1,39 @@
+        operationId: deleteUser
+        action: send
+        dependsOn:
+          - $workflows.dependsOnWorkflow.steps.step2
```
Could you please update the schema?
Because now [description said](https://github.com/OAI/Arazzo-Specification/pull/367/changes#diff-1e714ffe4a88e9eada463c25ca80ec1f0474ae43bbebc753bc31c4a240533a2bR182), that we MUST use only step in current workflow.

nashjain replied:
To avoid confusion, I've dropped the fully qualified step path from this file

AlbinaBlazhko17 commented on src/schemas/validation/schema.yaml:
```diff
@@ -240,21 +256,87 @@ $defs:
+      - $ref: '#/$defs/step-object-base'
+      - type: object
+        properties:
+          workflowId:
```
I validated this schema and found, that we can use `operationId` and `workflowId` in the same schema simultaneously. Could you please fix this behavior?

nashjain replied:
Can you please share the sample? In `tests/schema/fail/invalid-workflow-step-with-operationId.arazzo.yaml` I've demonstrated that both are not possible in the same spec.

AlbinaBlazhko17 replied:
Yes, sure. I am using this service to validate the schema: https://www.jsonschemavalidator.net/.
I put the schema to validate, which is simplified version of what we have now:
``` json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "oneOf": [
    {
      "allOf": [
        {
          "properties": {
            "stepId": {
              "type": "string"
            },
            "operationId": {
              "type": "string"
            }
          },
          "required": [
            "stepId",
            "operationId"
          ]
        }
      ],
      "unevaluatedProperties": false
    },
    {
      "allOf": [
        {
          "properties": {
            "stepId": {
              "type": "string"
            },
            "operationId": {
              "type": "string"
            }
          },
          "required": [
            "stepId"
          ]
        },
        {
          "properties": {
            "workflowId": {
              "type": "string"
            }
          },
          "required": [
            "workflowId"
          ]
        }
      ],
      "unevaluatedProperties": false
    }
  ]
}
```

And passed to it the example:
```json
{
  "stepId": "someStep",
  "workflowId": "someWorkflow",
  "operationId": "someOperation"
}
```
With this case the example with both `workflowId` and `operationId` is valid.

frankkilcommins replied:
Hi @nashjain, yes @AlbinaBlazhko17 is correct here in so far as the `workflow-step-object` within the JSON Schema does not check for mutual exclusivity with fields like `operationId`, `operationPath`, `channelPath`.

The sample failure Arazzo document used to test, namely https://raw.githubusercontent.com/specmatic/Arazzo-Specification/refs/heads/dev/tests/schema/fail/invalid-workflow-step-with-operationId.arazzo.yaml, is actually currently failing due to missing `sourceDescriptions`

Possibly you'd add a NOT constraint within the `workflow-step-object`

```
    not:
      anyOf:
        - required: [operationId]
        - required: [operationPath]
        - required: [channelPath]
```

nashjain replied:
Thanks @AlbinaBlazhko17 that was a very good catch. Sorry for the oversight. I've not pushed operationId to openapi-step-object and asyncAPI-step-object to avoid this issue. This also simplified a few other things. Tests are updated accordingly and passing. /cc @frankkilcommins

DmitryAnansky commented on src/schemas/validation/schema.yaml:
```diff
@@ -179,13 +179,13 @@ $defs:
         type: integer
       dependsOn:
-        description: Specifies a list of step identifiers that must complete (or be waited for) before the current step can begin execution. If you depend on a step from another workflow, you MUST use the full reference format with the workflow id.
+        description: Specifies a list of step identifiers that must complete (or be waited for) before the current step can begin execution. Steps referred by dependsOn SHOULD be non-blocking/async steps. If you depend on a step from another workflow, you MUST use the full reference format with the workflow id.
```
@frankkilcommins / @nashjain 
Also wondering how we defined `async step has completed`.
Should we wait for any message, first one or specific SuccessCriteria to pass?

nashjain replied:
Typically, an async dependency exists because a later step needs data from a message expected on a specific topic or channel.

So in such cases, you would depend on the output of a async receive step to complete within a timeout period. When we receive the expected message payload on the expected topic, we would consider it as complete. This is implicit and we don't need to define any SuccessCriteria.

DmitryAnansky replied:
How do you define the “expected message payload”?
From what I understand, we can only subscribe to a channel (i.e., receive an AsyncAPI response), but the exact shape or type of the expected message isn’t always strictly enforced.

Is it possible for an error message to be received when we’re expecting a specific receive message?

I can imagine a scenario where the incoming AsyncAPI message does not match what we expect. As a result, our output might not be properly resolved. In cases where there is no successCriteria (or another explicit way to mark a step as failed), the tooling doesn’t have a clear way to determine that something went wrong.

This leads to a few issues:
- A step failure may not be detected immediately.
- Outputs could end up being undefined without any clear signal.
- The problem might only surface later, when those outputs are used in subsequent steps.

How should such situations be handled?

tatomyr replied:
> Typically, an async dependency exists because a later step needs data from a message expected on a specific topic or channel.

Thanks @nashjain. So, is the only purpose of `dependsOn` at the step level to indicate that we use outputs from a previous step (or wait for that step's side effect), rather than to control parallelism (as it does at the workflow level)? What prevents us from setting outputs in this step and using them in subsequent steps without `dependsOn` being explicitly set? Do we actually need `dependsOn` at the step level, then? My take is that we should avoid adding things to the spec that aren't strictly necessary. Moreover, introducing something similar to an existing concept but with different behaviour feels a bit confusing.

nashjain replied:
That's a great question @tatomyr. 

If you see my original PR, we had a fork/join semantics. You would use a `fork` to allow an async step to start work that continues independently of the current linear step execution, and a later step can declare `join` to mean “do not proceed until that async work has reached its completion point.”

It was proposed that the spec already has a `dependsOn` to sever a similar purpose for workflows and we should use that instead of introducing a new keyword. We agreed and changed it to `dependsOn`

So at the step level, dependsOn is not only about consuming outputs from a previous step. In a purely sequential model, that would indeed be redundant. Its value is when an async step represents in-flight work, such as waiting for a message/callback/event, and another step needs to explicitly join on that work before continuing.

Hope this helps.

frankkilcommins replied:
> How do you define the “expected message payload”?
> From what I understand, we can only subscribe to a channel (i.e., receive an AsyncAPI response), but the exact shape or type of the expected message isn’t always strictly enforced.
>
>Is it possible for an error message to be received when we’re expecting a specific receive message?
>
>I can imagine a scenario where the incoming AsyncAPI message does not match what we expect. As a result, our output might not be properly resolved. In cases where there is no successCriteria (or another explicit way to mark a step as failed), the tooling doesn’t have a clear way to determine that something went wrong.
>
>This leads to a few issues:
>
>- A step failure may not be detected immediately.
>- Outputs could end up being undefined without any clear signal.
>- The problem might only surface later, when those outputs are used in subsequent steps.
>
>How should such situations be handled?

@DmitryAnansky regarding your points above, they are valid concerns. I think we can address ambiguity by adding the following clarifying section to the spec:

```text
##### Defining Success for Asynchronous Steps

For steps that reference AsyncAPI operations (via `operationId` or `channelPath`), tools will send or receive messages on the specified channel as defined by the referenced AsyncAPI description. For AsyncAPI send operations, the step completes immediately after the message is sent. Arazzo does not model broker acknowledgment or delivery confirmation. For AsyncAPI receive operations, step completion depends on message receipt as described below.

Because AsyncAPI channels MAY define multiple message types, and because message payloads MAY represent either success or failure conditions, authors SHOULD define `successCriteria` for AsyncAPI receive steps to explicitly evaluate the received message payload (for example, via `$message.payload`) and determine whether the step succeeded.

Authors MAY omit `successCriteria` only when both of the following conditions are met:

- The channel defines a single message type that unambiguously represents successful completion
- The message payload does not contain fields indicating error states (e.g., status codes or error flags)

When `successCriteria` is omitted, receiving any message matching the `correlationId` (if specified) within the `timeout` period is considered a successful step completion.

If `correlationId` is specified, only messages matching the correlation identifier are considered. If no matching message is received within the `timeout` period, the step fails and triggers any defined `onFailure` actions.
```

frankkilcommins replied:
@tatomyr @nashjain 

Regarding comments here on step level `dependsOn` and in light of the older comment thread on the topic (see https://github.com/OAI/Arazzo-Specification/pull/367#discussion_r2671672797), I would propose the following section be added to add clarity and remove ambiguity around what step level dependsOn introduces and opens the possibility too. Whether or not tooling chooses to support all aspects is a tooling choice, but I propose some clarity in that respect too to make it easier for tooling authors to follow).

```text
##### Step Dependencies and Execution Order

The `dependsOn` field at the step level is primarily intended to coordinate asynchronous operations.

**Recommended Approach for Synchronous Workflows**

For workflows containing only synchronous steps, the RECOMMENDED approach is to order steps sequentially in the steps array without using `dependsOn`. This provides the simplest and clearest execution model. Step-level `dependsOn` is typically unnecessary when all operations complete synchronously and execution follows a linear path.

**Use Case: Async Coordination**

When a step must wait for an asynchronous operation to complete before proceeding, `dependsOn` establishes a join point for in-flight async work. For example, a step that requests an order status SHOULD declare `dependsOn` on the step that receives order creation status from an async order placement, even if no explicit output reference exists. This ensures the async operation completes before the dependent step executes.

**Authoring Guidance**

For async workflows, authors SHOULD use `dependsOn` to explicitly declare when a step must wait for async work to complete, regardless of output references. This is the intended use case for step-level dependencies.

**Tool Behavior**

Tools MUST respect all declared `dependsOn` relationships. Tools MUST also treat runtime expression output references (e.g., `$steps.stepId.outputs.field`) as implicit dependencies and ensure the referenced step completes before the referencing step executes.

Tools supporting only sequential execution MUST execute steps in an order that satisfies both explicit (`dependsOn`) and implicit (output reference) dependencies.

**Validation Recommendations**

Implementations SHOULD validate the following scenario:

When no `dependsOn` is used in a workflow (sequential execution model), implementations SHOULD produce an error if a step references another step's outputs where the referenced step appears later in the steps array. This indicates a forward reference that cannot be satisfied in sequential execution.
```

In the interest of continuous improvement and moving the spec forward, if a section like this, and also like the one from the previous comment to clear up step success for async steps are added to Arazzo, then I am more that happy to move forward with including this PR into the v1.1-dev branch and preparing for the 1.1 release.

nashjain replied:
The PR is updated.

nashjain committed:
fix: remove fully qualified dependsOn field from async-depends-on.yaml

nashjain committed:
fix: refactor step schema to isolate operation-specific fields and deduplicate operation step parameter rules

nashjain committed:
Merge branch 'v1.1-dev' into dev

frankkilcommins commented on examples/1.1.0/pet-asyncapi.yaml:
@nashjain not sure if this file just popped in now as part of the additional merge work (or slipped through initial review), but I don't see the need to create it within the src folder. We really only want the spec md and the schemas there.

If you want to keep this file then it could go in `/examples/1.1` along with an Arazzo 1.1 Document that leverages it. Otherwise, just delete from this PR and we can handle having some 1.1 examples as part of #467

nashjain replied:
this file was always there since it is referred from arazzo.md file. I've moved it to examples folder.

nashjain committed:
chore: move pet-asyncapi.yaml to examples folder

nashjain committed:
docs: clarify async workflow step dependencies and success criteria

nashjain committed:
docs: fix markdown headings

nashjain committed:
docs: clarify async step dependencies and completion semantics

nashjain committed:
docs: clarify async step dependencies and completion semantics

frankkilcommins reviewed (approved):
Great work @nashjain, and thanks for the additional clarification text around step `dependsOn`, and determining success for async steps.
